cordapp-example/workflows-kotlin/src/main/kotlin/com/example/flow/PaybackFlow.kt

```diff
@@ -0,0 +1,160 @@
+            val txCommand = Command(IOUContract.Commands.Destroy(), inputIOUState.participants.map { it.owningKey })
+
+            // Obtain a reference to the notary that was in use for input state
+            val notary = results.states[0].state.notary
+
+            // build tx that will consume previous output
+            val txBuilder = TransactionBuilder(notary)
+                    .addInputState(inputStateAndRef)
+                    .addCommand(txCommand)
+
+            progressTracker.currentStep = VERIFYING_TRANSACTION
+            // Verify that the transaction is valid.
+            txBuilder.verify(serviceHub)
+
+            progressTracker.currentStep = SIGNING_TRANSACTION
+            // Sign the transaction (must be by a borrower)
+            val signedTx = serviceHub.signInitialTransaction(txBuilder)
+
+            progressTracker.currentStep = INITIALIZE_OTHER_PARTY_FLOW
+            // Send the state to the counterparty, and receive it back with their signature.
+            val lenderPartySession = initiateFlow(inputIOUState.lender)
+
+            val confirmPayback = lenderPartySession
+                    .sendAndReceive<Boolean>("ready for receiving payback?")
+                    .unwrap { it }
+
+            require(confirmPayback) { "Oops, lender failed to be ready" }
+
+            progressTracker.currentStep = PAYBACK_COMMUNICATION
+            lenderPartySession.send(paybackValue)
+
+            progressTracker.currentStep = GATHERING_SIGS
+            val fullySignedTx = subFlow(CollectSignaturesFlow(
+                    signedTx,
+                    setOf(lenderPartySession),
+                    ExampleFlow.Initiator.Companion.GATHERING_SIGS.childProgressTracker())
+            )
+
+            progressTracker.currentStep = FINALISING_TRANSACTION
+            // Notarise and record the transaction in both parties' vaults.
+            return subFlow(FinalityFlow(fullySignedTx, setOf(lenderPartySession), FINALISING_TRANSACTION.childProgressTracker()))
+        }
+    }
+
+    @InitiatedBy(Initiator::class)
+    class Acceptor(val borrowerPartySession: FlowSession) : FlowLogic<SignedTransaction>() {
+        @Suspendable
+        override fun call(): SignedTransaction {
+            borrowerPartySession.receive<String>().unwrap { it }
+            //send confirmation about being ready to receive payback
+            borrowerPartySession.send(true)
+
+            val receivedPayback: Int = borrowerPartySession.receive<Int>().unwrap { it }
+
+            val signTransactionFlow = object : SignTransactionFlow(borrowerPartySession) {
+                override fun checkTransaction(stx: SignedTransaction) = requireThat {
+                    val stateAndRef = serviceHub.toStateAndRef<ContractState>(stx.tx.inputs[0])
+                    val expectedAmount = (stateAndRef.state.data as IOUState).value
+
+                    "Payback value differ from borrowed amount." using (receivedPayback == expectedAmount) // */
```

Comment:
This `receivedPayback` value seems contrived. With this `==` you seem to "not trust" the initiating party, but you "trust" the initiating party with the value of `receivedPayback`... Again, other than as a learning experiment, I don't see the point.
On the other hand, what the responder could do is confirm that it is an IOU to which it is a party.

Reply:
yes, just experiment.

> On the other hand, what the responder could do is confirm that it is an IOU to which it is a party.

Isn't this done in contract, when I require sig of both participants? if responder is not a party, tx will be rejected, right?

Reply:
Imagine this case:

* Alice (lender) creates IOU with Bob (borrower)
* Alice starts flow to destroy, asks Bob and Carly to sign.
* Bob and Carly sign
* The contract is happy because Alice and Bob signed

Why did Carly sign?? Sounds like funny business, why would you ask a signature from a party that does not need to be a signer? Perhaps an attack is under way...

Reply:
Alice can sub-class the initiating flow to do more things and Bob and Carol only know it comes from the initiating flow.

Reply:
Now I got it.